Apiserver namespace labels

jayunit100 opened the pull request:
thx to @rikatz @thockin , ok ! here we go... lets see if we can workaround the issues so far in https://github.com/kubernetes/enhancements/pull/2113/files by adding default labels.

Fixes #2161

jayunit100 commented:
/assign @liggitt

jayunit100 commented:
/assign @thockin

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
````diff
@@ -0,0 +1,408 @@
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+if obj.Labels == nil {
````
missing indent :)

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
````diff
@@ -0,0 +1,408 @@
+
+```
+func ValidateNamespace(namespace *core.Namespace) field.ErrorList {
+	// ...
````
weird indents

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+This solves not just NetworkPolicy, but all such namespace selector APIs without any disruption or version skew problems.
+
+### Test Plan
+
```
We need a plan here for testing - we can do a unit test that loops over namespaces, and does a select by this label and verifies the results, plus the obvious tests for trying to modify it and failing

thockin replied:
defaults.go populates on read, but I don't know about watch.  We need to test that WATCH on objects that already exist and DO NOT have the label actually written will work, and if not we need to find a plan

jayunit100 replied:
why would a watch *not* work?  Its just the mutation of a label, right ? 
I suppose someone could write a watch that selected `! kubernetes.io/metadata.name` but that would 
be a very odd circumstance , i.e. unless someone went into the future and wanted to avoid auto labelled namespaces after reading this KEP ? :) 

or i might be missing a more obvious breaking scenario?

thockin replied:
I thought maybe it would not work, but I just tried it and watch does seem to work correctly.

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
```
1.21:
gate disabled by default
if gate disabled, log or event incompatible uses of this
if gate enabled, enforce
release note

1.22:
gate enabled by default
release note

1.23:
lock gate enabled

jayunit100 replied:
added

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+
+
+### Upgrade / Downgrade Strategy
+
```
Need to check the watch issue above.  If that works, great.  If not, we may need to add a sweeper that populates the label at startup.

thockin replied:
In case of downgrade, any policies that rely on this will either work fine (if the label was written) or fail-closed becasuse the label does not exist.  Acceptable.

jayunit100 replied:
Since reads go through `defaults.go` you get lazy population in watches, so this works as intended

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+
+
+### Version Skew Strategy
+
```
NA

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
```
It's kind of a special hack.  The fact that it isn't generalized or generalizable makes it feel special.

jayunit100 replied:
fixed

jayunit100 replied:
we could parameterize this, right ? `--namespace-default-label-prefix` or something?  not super valuable but it could be parameterize

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,29 @@
+reviewers:
+  - @liggitt
+  - @nikhita
+approvers:
```
@liggitt willing to serve as approver?  Since I am too close to it?

rikatz replied:
Oh sorry, changed wrongly. I'll put @liggitt as approver then

thockin replied:
We should ask him :)

jayunit100 replied:
Yup Asked in slack he said he'll check it out tomorrow.

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,29 @@
+latest-milestone: "v1.21"
+milestone:
+  alpha: "v1.21"
+  beta: "v1.21"
```
22

jayunit100 replied:
fixed

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,29 @@
+milestone:
+  alpha: "v1.21"
+  beta: "v1.21"
+  stable: "v1.21"
```
23

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,29 @@
+  - none
+disable-supported: true
+metrics:
+  - my_feature_metric
```
none

jayunit100 replied:
fixed

rikatz commented:
Made some suggested corrections.

The forcepush / squash was only because I've made some mess with kep.yaml...

jayunit100 commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+Both of these are impractical due to the nature of security tools, which need to be robust and explicit.
+
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy (which are out-of-tree - generally implemented by cni plugins) and would not introduce new surprises for users… But in practicality this is not the case because, by adding a new selection data structure, the logical inclusion/exclusion properties of namespace and pod selectors become drastically different, as sh ...
```
added

jayunit100 commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+- closed failures which break on old clients (making networkpolicy api evolution very hard)
+- open failures with lots of warnings and docs about how old clients might overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to be robust and explicit.
```
ok, added a statement to this effect

jayunit100 commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+### Non-Goals
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in k/k - Validating and Mutating webhooks.  They will inevitably need similar API evolution, but we don't  need to do that as part of this KEP.
+
```
updated by @rikatz

jayunit100 commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+
+### Non-Goals
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in k/k - Validating and Mutating webhooks.  They will inevitably need similar API evolution, but we don't  need to do that as part of this KEP.
```
fixed

jayunit100 commented:
cool thanks ricardo.  added the table that was deleted back.

jayunit100 commented:
@andrewsykim FYI ! here it is :)

thockin reviewed (commented):
I am overall happy - still a number of sections to fill out and boilerplate to remove :)

jayunit100 commented:
will get boilerplate cleaned up later today, thx again everyone.

thockin commented:
How many years of "why can't we" will be solved by 12 LOC.

abhiraut commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,368 @@
+
+If this label is missing, it is added by the apiserver
+
+If the label is incorrect or changed, the apiserver fails to validate the object
```
this is probably a remote and maybe even unlikely possibility, what happens to namespaces, prior to upgrade, who had same label key(whatever we choose) but a different value.. these namespaces will now be invalidated which previously worked fine.

andrewsykim replied:
Good question -- we own `kubernetes.io/` prefixed labels so I don't think we are violating any compatibility rules if a user was using it for some other use-case outside of what Kubernetes is managing.

abhiraut replied:
thanks Andrew. I am aware of this. However we don't validate against it so it remains a possibility. Only reason i raised this point is to clarify the approach that will be taken for such a case.

andrewsykim replied:
Good point -- we should at least write down what the expected behavior for that user would be. It sounds like worse case scenarios is we disallow any future writes to that namespace object.

thockin replied:
This is why we should have one alpha release where we warn if we see people using it

jayunit100 replied:
seems like we should , in general make this  label EXTREMELY specific so that this just **doesnt** happen 
  `k8s.io/apimachinery.namespace.default.name` (alternatively , literally plumb it in as an apiserver option that can be overriden if folks want to)

abhiraut replied:
i think the name proposed in the KEP is perfectly fine IMO..

jayunit100 replied:
Ok happy to keep the name as is , resolving

abhiraut replied:
updated text

thockin replied:
First question is whether we want to fail if someone tries to set the value.  That was my initial feeling, but it sounds like you are going full-tilt.  I'm OK with that, I guess.  It does answer "what about pre-existing".

The simple answer seems to be to just blindly set the value.  I think defaults is good enough, isn't it?  The only downsides is that it is versioned.  Doing it in PrepareFor and Decorator is more involved.  Can I select on labels that only are set in Decorator?

liggitt replied:
> First question is whether we want to fail if someone tries to set the value. That was my initial feeling, but it sounds like you are going full-tilt. I'm OK with that, I guess. It does answer "what about pre-existing".

I like my invariants invariant

> The simple answer seems to be to just blindly set the value. I think defaults is good enough, isn't it?

Though it's unorthodox for defaults to stomp existing values, that would certainly be the simplest place to do it (done in a single place, applies to reading existing data and accepting create/update requests). We could go belt-and-suspenders with a validation check that requires the label exist and match the name (so if we forget to add defaulting for "v2" namespaces it would be caught).

The only thing stomping in defaults would not let us do would be to detect/warn if a user tried to set the label to some other value (since by the time we got to the strategy the conflicting user-specified value would be gone). That seems unlikely to be done accidentally, so I'm not sure how much I care about that.

abhiraut replied:
from what i gather, both of you sound _OKAY_ with "default stomping" except that it does not allow us a chance to warn users, but coupled with the fact that the label actually belongs to a reserved prefix of "kubernetes.io/", its okay to override such (ill-advised) usage of these labels?

evankanderson replied:
Ha, I had two thoughts here:

1. After this feature is enabled, will label selectors work for Namespaces, or will there need to be a read-write cycle on each Namespace object (like with CRDs for version upgrades)?

2. Would it make sense to use/reserve `metadata.kubernetes.io/` as a prefix for this kind of usage (reflecting field values to labels) without any promises of future reflections?

liggitt replied:
> 1\. After this feature is enabled, will label selectors work for Namespaces, or will there need to be a read-write cycle on each Namespace object (like with CRDs for version upgrades)?

this would work on read, no write cycle required

> 2\. Would it make sense to use/reserve `metadata.kubernetes.io/` as a prefix for this kind of usage (reflecting field values to labels) without any promises of future reflections?

I'd tend to keep the `metadata.name` construct together rather than splitting it

jayunit100 replied:
I LOVE idea # 2 because @deads2k gently suggested that, why not reflect other scalar api fields

jayunit100 replied:
this thread is long: oversimplifying the comments:  


```
@thockin: why we should have an alpha release
@liggit: reasonable to make use of kube.io labels w/o an intermediate release
         strongly force the label value
          - add if absent
          - prepareForCreate/PrepareForUpdate ~ override wrong values
          - Decorate: replace differeing values
@thockin is ok with ^, but "Can I select on labels that only are set in Decorator?"
@evan reserve `metadata label` + also wants "add if absent"

```

thockin replied:
Re name: <something.>kubernetes.io/metadata.name seems best - extends well and maps to field-selector syntax.  The sentiment on "<something.>" seems to be "", which I am fine with.

Re alpha: I am not convinced - I still think a cycle of warning is "the polite thing to do".  After that, I am OK with unilateral stomping.

Re where: I tested defaults and it seems to work.  I tested Decorator and it does NOT seem to work.  This actually seems like a problem to me, but not one for today.

thockin commented:
We should not adulterate the API that we present artificially.  It's
already an INCREDIBLY precise name proposal, I think.

On Mon, Nov 23, 2020 at 12:18 PM jay vyas <notifications@github.com> wrote:

> *@jayunit100* commented on this pull request.
> ------------------------------
>
> In keps/sig-api-machinery/2161-apiserver-default-labels/README.md
> <https://github.com/kubernetes/enhancements/pull/2162#discussion_r528971428>
> :
>
> > +- Add an immutable default label to ALL namespaces, as the namespace name so this
>
> +can be used as selector by arbitrary components.
>
> +
>
> +### Non-Goals
>
> +
>
> +* Publishing name labels on all resources
>
> +* Exposing arbitrary fields as selectors
>
> +
>
> +## Proposal
>
> +
>
> +We propose to label of all namespaces, by default, with a reserved label
>
> +(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
>
> +
>
> +If this label is missing, it is added by the apiserver
>
> +
>
> +If the label is incorrect or changed, the apiserver fails to validate the object
>
>
> seems like we should , in general make this label EXTREMELY specific so
> that this just *doesnt* happen
> k8s.io/apimachinery.namespace.default.name (alternatively , literally
> plumb it in as an apiserver option that can be overriden if folks want to)
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/2162#discussion_r528971428>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVAYXUX56HFVJJ76N73SRK7RZANCNFSM4T6RGZIQ>
> .
>

thockin commented:
We should figure out who feels they "own" namespaces logic and who feels
like they would want input on this.  I'll ping sig-arch

On Tue, Nov 24, 2020 at 4:51 AM jay vyas <notifications@github.com> wrote:

> *@jayunit100* commented on this pull request.
> ------------------------------
>
> In keps/sig-api-machinery/2161-apiserver-default-labels/README.md
> <https://github.com/kubernetes/enhancements/pull/2162#discussion_r529521033>
> :
>
> > +- Add an immutable default label to ALL namespaces, as the namespace name so this
>
> +can be used as selector by arbitrary components.
>
> +
>
> +### Non-Goals
>
> +
>
> +* Publishing name labels on all resources
>
> +* Exposing arbitrary fields as selectors
>
> +
>
> +## Proposal
>
> +
>
> +We propose to label of all namespaces, by default, with a reserved label
>
> +(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
>
> +
>
> +If this label is missing, it is added by the apiserver
>
> +
>
> +If the label is incorrect or changed, the apiserver fails to validate the object
>
>
> Ok happy to keep the name as is , resolving
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/enhancements/pull/2162#discussion_r529521033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVCQHKP354QMIP3ECQ3SROT5NANCNFSM4T6RGZIQ>
> .
>

robscott commented:
Thanks for this KEP @jayunit100! This would be a huge win for [service-apis](https://github.com/kubernetes-sigs/service-apis). We've been heavily reliant on selectors and had been thinking we'd eventually have to add support for a list of namespace names in some places to get around the limitations of selectors.

robscott commented:
/cc

jayunit100 commented:
Thanks mainly to  @thockin for really getting this concept figured  out :)  wouldn't have had any idea to start otherwise.

jayunit100 commented:
Pushed some boilerplate cleanup

Liggit said he'll check on Wednesday. So we should have some daylight by then to see what happens next @robscott .  If anything else to discuss feel free to ping on slack will be around today and tomorrow .

Hope to have this "stable" by end of "week".

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is reserved. However, since there
```
Find the link to where we reserved them all.  These are pretty clearly "ours" and I remember writing a doc claiming them and a large chuck of namespaces at the same time.

abhiraut replied:
will update the doc with this link https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ which explicitly calls this out.

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+  	obj.Labels = map[string]string{}
+  }
+  
+  if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
```
seems like you would want this unconditional.  if this isn't forced, then your selection isn't guaranteed.

liggitt replied:
I made a similar comment about forcing the invariant in https://github.com/kubernetes/enhancements/pull/2162/files#r530463637

One benefit to overriding in strategy is that a warning can be added indicating the label on the incoming request was overridden.

jayunit100 replied:
@liggitt  Assuming, qouting you from above :), that since the invariant can be forced on read anyways, just that we lose warnings, this is resolved ... 
```
The only thing stomping in defaults would not let us do would be to detect/warn if a user tried to set the label to some other value (since by the time we got to the strategy the conflicting user-specified value would be gone). That seems unlikely to be done accidentally, ....  not sure how much I care about that.
```

@deads2k if you feel very strongly that we **must** warn users about this, then LMK and we can revisit using strategy.go I think.  

Thanks!

thockin replied:
I think we can do:

defaults.go:
```
if gate-enabled {
    if obj.Labels == nil {
￼  	    obj.Labels = map[string]string{}
￼    }
    obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
}
```

Then, elsewhere (@liggitt - where can we send API warnings?):

```
if not gate-enabled {
    if obj,Labels != nil && obj.Labels[“kubernetes.io/metadata.name”] != obj.Name {
        API warning "the label 'kubernetes.io/metadata.name' is reserved and such uses will be over-written starting in v1.22"
    }
}
```

Is that fair?  Is it too generous?  It's unlikely to REALLY be an issue, but we just can't know...

@liggitt Remind me why it's better to just stomp the value rather than fail the API operation if a user tries to change it?

liggitt replied:
> @liggitt - where can we send API warnings?

we can send them from strategy methods (PrepareForCreate / PrepareForUpdate / Validate / ValidateUpdate) and anything we plumb the request context to

> Is that fair? Is it too generous? It's unlikely to REALLY be an issue, but we just can't know...

I think that's too generous. Setting a reserved label with a pretty clear name (in my mind, at least) to a value that does not match the label does not seem like a reasonable thing to delay a release over.

> @liggitt Remind me why it's better to just stomp the value rather than fail the API operation if a user tries to change it?

Failing the request means:
1. the failure is not API-visible (since the write is disallowed)
2. we're depending on the rejected caller behaving well in response to the rejection (not hot-looping, reporting the failure in a discoverable way)
3. fixing the issue means fixing the caller, which could involve getting a new copy of some operator from the provider (the owner of the cluster where the call is rejected might not be the owner of the problematic writer).

In contrast, stomping the label has the same end-result of a controller that updates the `kubernetes.io/metadata.name` label asynchronously, which I think would be a fair thing to do (since we reserve ownership of kubernetes.io labels). There's still the hypothetical possibility of a component that was 1) setting this kubernetes.io label differently and 2) behaving poorly when the value later did not match it's set value, but given how specific the label name is, I do not expect that would happen accidentally.

jayunit100 replied:
i think this is all set ?   we generally agree that we'll stomp.

liggitt replied:
Re: precedent, this seems similar when we picked GA `kubernetes.io/os` and `kubernetes.io/arch` label keys, and started reconciling them for Nodes. We didn't sweep for user-set labels with different values and take multiple releases to phase in the new ones, allowing conflicting user-set values in the meantime. I think the `kubernetes.io` label prefix gives us reasonable leeway here.

> I do not feel like waiting one release is a particular hardship, given this has been unsolved for literally years.
> ...
> OK. I buy that argument.

I can't tell whether you still want to phase it in over multiple releases or not.

jayunit100 replied:
i feel like we can decide to wait a release cycle, or not, irrespective of this specific PR, so ill mark this as resolved, and then follow on w / @thockin in the next sig-network meeting (tomorrow)

thockin replied:
I spoke to @liggitt offline and I am conceding.  I think we do not NEED to do a cycle of warnings.  We can just enable this.

Given the scope, we can probably just go straight to GA. ?

deads2k commented:
> We should figure out who feels they "own" namespaces logic and who feels like they would want input on this. I'll ping sig-arch

Given that it's the grouping primitive for the kube-apiserver I would expect sig-apimachinery.  We've also talked about namespace related topics like initialization.

/cc @lavalamp

deads2k commented:
Looking over the proposal, you're trying to add a label selector that could only be used for `=`, `==`, `in` operations?  Since we don't support `OR` expressions or patterns in label selection, what is this buying you over a GET or a LIST with a field selector?  Do users really prefer writing the label selector over specifying a list of namespaces?

deads2k commented:
> Looking over the proposal, you're trying to add a label selector that could only be used for `=`, `==`, `in` operations?

I think I could see `not in` being a differentiator in favor of creating a label like this.

Could you try to reduce the size of motivation to something more end user API focused and less focused on why you think a default label is the right solution?  The information there could be in a different section about "why label selector"

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+possible to deal with labels.
+
+The idea of this KEP is to propose that namespaces have an immutable label when created, created
+by the APIServer that represents the metadata.name of the object and turn namespaces selectable 
```
Why would namespaces be special in this regard as opposed to all resources?  I'm not asking for a change, I'm asking for a reason why a system maintained label would not be consistent across all resources.

jayunit100 replied:
So, arbitrary labels on everything would be cool, but scalability issues possibly make it out of scope for thisproposal... . imagine a deployment that scaled pods to 100 or 1000 replicas.  the case you'd want to make to add a gauranteed  30 byte label there, might be different then that of namespaces, which are generally sparse and less dynamic.

liggitt commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+### Upgrade / Downgrade Strategy
+* Check the WATCH situation described in test plan, otherwise populate the label 
+of namespaces without it.
+* In case of downgrade, as this is just a label it will work fine. Network Policies 
+that relies on this label may fail-close, which might be acceptable.
```
> Network Policies that relies on this label may fail-close, which might be acceptable.

The may/might is squishy... it would be good to pin down that language.

Also note that selectors can be used in an *exclusive* way to try to allow access everywhere *except* certain namespaces:

```yaml
namespaceSelector:
  matchExpressions:
  - key: kubernetes.io/metadata.name
    operator: NotIn
    values:
    - kube-system
```

if the namespace is missing the label, that selector would also match the namespaces it was expected to exclude, right? That's a sort of strange construction for network policy, but I think that would be more likely to happen in a webhook admission configuration that was attempting to avoid running against `kube-system`, and accidentally would on older versions lacking the label on that namespace.

I don't really see a way around that, but it's worth mentioning in the docs ("when used on releases prior to X, namespaces are not guaranteed to have this label").

liggitt commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
```
One of the interesting things about this proposal is that it can be back-filled out-of-tree on older releases with a simple namespace-labeling controller... would it be valuable to build/provide that as well?

jayunit100 replied:
@abhiraut was thinking to own this via an admission controller / web hook I believe

liggitt replied:
a reconciling controller component to fixup namespaces created prior to existence of the admission plugin would be needed as well

evankanderson replied:
Would this need to be a controller or a one-shot job?

jayunit100 replied:
need to clarify what we mean by 'on read' above, will look into this now

jayunit100 replied:
it seems like if folks wanted to back fill we could easily port this into patch releases ?   After thinking about it more it feels like a controller might be alot of engineering for not much benefit

liggitt replied:
> we could easily port this into patch releases ?

As a policy, we don't backport features to patch releases.

> After thinking about it more it feels like a controller might be alot of engineering for not much benefit

Depending on the environment, accelerating the ability to use these selectors against older releases could be very useful.

jayunit100 replied:
ok, do you want that to be part of this KEP or a community effort that is complimentary ?

thockin replied:
That's outside the scope of this.

jayunit100 replied:
ok resolving, we can discuss the controller idea in a different forum

liggitt commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+
+### Feature Enablement and Rollback
+
+1.21:
+- gate disabled by default
+- if gate disabled, log or event incompatible uses of this
+- if gate enabled, enforce
+- release note
+
+1.22:
+- gate enabled by default
+- release note
```
if we're looking for feedback from consumers, could we gain that earlier/more quickly with a namespace-labeling controller and accelerate this?

jayunit100 replied:
in general most of the NetworkPolicy APIs could be prototyped in the early stages via some type of lightweight CRD or controller .  I think we should visit this monday .  cc @rikatz @abhiraut ha look the netpol operator idea ... just ... won't. .. go ... away  :)

jayunit100 replied:
resolving , assuming that the controller feedback idea isnt an actual blocker for this as a feature, but more an idea we can consider

liggitt commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
```
suggest a concrete proposal here:

```suggestion
We propose to label of all namespaces, by default, with a reserved label 
of `kubernetes.io/metadata.name` to allow easy selection.
```

liggitt commented:
I'm generally in favor of this, though I agree with Tim that it's kind of a special hack. Being clear about why this is scoped to namespaces (https://github.com/kubernetes/enhancements/pull/2162#discussion_r530418258) would be good. One concrete reason is that not all objects' names are valid label values.

> Looking over the proposal, you're trying to add a label selector that could only be used for `=`, `==`, `in` operations? Since we don't support `OR` expressions or patterns in label selection, what is this buying you over a GET or a LIST with a field selector? Do users really prefer writing the label selector over specifying a list of namespaces?

A few things won me over:
* the in/not-in constructions seem useful
* enabling all in- and out-of-tree policy mechanisms that currently support namespace label selection to gain this capability with no changes on their part
* it side-steps the issues with a policy being more or less permissive than intended when evaluated by clients that aren't aware of the new fields in the policy (especially important for networkpolicy which is enforced out of tree)

jayunit100 commented:
> Looking over the proposal, you're trying to add a label selector that could only be used for `=`, `==`, `in` operations? Since we don't support `OR` expressions or patterns in label selection, what is this buying you over a GET or a LIST with a field selector? Do users really prefer writing the label selector over specifying a list of namespaces?

Well, its really more for users wanting specifically to target a namespace, in a networkpolicy, by name.  You could select things based on field for other types of operations if you had the whole go client ...  but the networkpolicy API is not quite that flexible, and is inherently designed to select on Labels, specifically, not arbitrary fields.   Since CNI providers are quite diverse, telling all of them to use the go client to support such an operation is impracticle in general.

derekwaynecarr commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
```
As part of existing namespace admission controller?  Assume we will have a loop in controller manager that applies it for existing namespaces if not present?

jayunit100 replied:
@abhiraut ^

derekwaynecarr replied:
ignore question about admission, but assume we will want some path to apply label after upgrade.  maybe kube storage version migration is sufficient.

jayunit100 replied:
Great catch .  Talking with @andrewsykim and @rikatz about this.  We'll add details to  the upgrade/downgrade section and disambiguate this .

jayunit100 replied:
updated to reflect the new thinking around defaulting APIs

derekwaynecarr commented:
I am +1 on the concept, just had a couple questions beyond what @deads2k and @liggitt identified.

liggitt commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,31 @@
+  - "@nikhita"
+  - "@thockin"
+approvers:
+  - "@liggitt"
```
someone from api-machinery as an approver would make sense as well... @deads2k is already reviewing, might make sense

jayunit100 commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,350 @@
+that relies on this label may fail-close, which might be acceptable.
+
+### Version Skew Strategy
+N/A
```
ack, yep, hadn't thought about the upgrade issues before (or maybe we just glanced over them) .  we'll update that.

thockin replied:
I presume that any policy which USES this label which is created DURING the control-plane update is OK to not work as expected.  More likely people will upgrade apiserver and then create policies.  defaults.go should cover all cases, I think.

jayunit100 replied:
added

jayunit100 commented:
Hiya folks, just discussed reviews from today w/ Ricardo and Andrew.   Will  update the KEP accordingly to have more granular info about how we're thinking about versions.

evankanderson commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,392 @@
+the existing API objects?**
+
+No
+
```
I think this will actually increase the resource size of Namespace objects and their label index (but probably not noticably).

abhiraut replied:
done

tallclair commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,391 @@
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this 
+can be used as selector by arbitrary components.
```
nit: This is more a summary of the implementation. I would say the goal is to enable things that select namespaces by labels can securely and reliably also select individual namespaces by name.

tallclair commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,391 @@
+
+### Non-Goals
+
+* Publishing name labels on all resources
+
+Not all resource names satisfy the label requirements. Some of those resources require
```
```suggestion
* Publishing name labels on all resources - Not all resource names satisfy the label requirements. Some of those resources require
```

(formats better in markdown)

tallclair commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,391 @@
+
+Not all resource names satisfy the label requirements. Some of those resources require
+a name that can be used as a DNS subdomain name.
+* Exposing arbitrary fields as selectors
```
```suggestion
* Exposing arbitrary fields to selectors
```

tallclair commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,391 @@
+“kubernetes.io/metadata.name” and set it to the namespace's name. This will allow
+any resource with a `namespaceSelector` field (such as NetworkPolicy, MutatingWebhookConfiguration
+etc) to select (or exclude with `notIn`) these namespaces by explicitly matching on
+the namespace's name with traditional label selector mechanisms.
```
Are we confident that all the namespace label selectors we care about use `metav1.LabelSelector` rather than `map[string]string`? The problem with the latter is that under this proposal it would only be able to match a single namespace at a time.

thockin replied:
All three are `metav1.LabelSelector`

abhiraut commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,408 @@
+We considered doing this for all resources, but some k8s resources allow object names that are not valid label values.  Given that we have a clear and concrete use-case in namespaces, it seemed proper to start here.  If other kinds need similar treatment (which seems unlikely) they can use the same label key if their name format allows.
+
+We considered telling users to DIY with Validating and Mutating admission controllers.  This is unpleasant at best and seems particularly egregious given that a) multiple APIs benefit from this; and b) the implementation is very simple.
+
```
added another section

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,438 @@
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
```
Can we start on the checklist to move this forward?

jayunit100 replied:
good point ! done

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,438 @@
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is [reserved](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). However, since there
+is no validation of reserved label prefixes, such namespaces will be invalidated
```
s/namespaces/labels?

jayunit100 replied:
fixed

thockin reviewed (commented):
Would love to press this forward and get it off the docket...

jayunit100 commented:
> Would love to press this forward and get it off the docket...

gotcha, yeah definitely ! i think were up to date on the comments and feedback now. 
@liggitt , @deads2k , @tallclair  do you have any outstanding issues ?

jayunit100 commented:
Re: closed failure vs. stomping: Should we make that final call in an iteration ? or in the actual PR https://github.com/kubernetes/enhancements/issues/2161 ?
That might be something that folks want to think through a little more.  the rest of the content here seems mostly agreed upon .

thockin reviewed (commented):
Would like to get this in and iterate on last niggles.

jayunit100 commented:
Hi @liggitt  ....  looks like we've resolved the outstanding comments , as i think the threads are mostly now at a point of `violent aggreement` :) .   

Roughly the summary i think we converged on is:
- warning release
- stomping is ok 
- labelling controller might be a good idea, not sure where it should live 
- we need to make sure we test watches 

Would we be able to  merge+iterate on the details be okay for this ?

liggitt commented:
I'm happy to merge and iterate. Per the KEP template:

> Any KEP marked as `provisional` is a working document and subject to change. You can denote sections that are under active debate as follows:
>
> ```
> <<[UNRESOLVED optional short context or usernames ]>>
> Stuff that is being argued.
> <<[/UNRESOLVED]>>
> ```

jayunit100 commented:
ok... we're in `provisional` status, so , i think we're go then for launch ?

jayunit100 commented:
/test

jayunit100 commented:
/test pull-enhancements-verify

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+
+1.21:
+- gate disabled by default
+- if gate disabled, log or event incompatible uses of this
```
@liggitt if we are unilaterally stomping, and the entire net change is 6 LOC + tests, do we even need an alpha cycle?

liggitt replied:
starting in beta, enabled by default, unilaterally stomping sounds good to me

jayunit100 replied:
@liggitt do we want to merge and iterate, or do you want me to make more changes now ... ?

liggitt replied:
I'm ok with a follow-up dropping the warning release, starting in beta, and resolving the unresolved sections before moving to implementable

/lgtm
/approve

(needs an api-machinery approver)

abhiraut replied:
ping @deads2k ^

thockin replied:
@liggitt what does "beta" mean for this?  Gate-disableable?

liggitt replied:
That was my thought, yeah. A one release grace period to opt out of the behavior at the cluster deployer level if needed in 1.21.

lavalamp replied:
gate-disableable for one release SGTM

jayunit100 replied:
pushed thanks ! yeah, removed 1.23

thockin reviewed (commented):
LGTM but I am not the approver and I left some near-resolution comments

jayunit100 commented:
Ok.... added UNRESOLVED tags so we can fine tune the stomping and design impl (validation vs preCreate vs....) argument as needed . 

Am assuming this PR is "done" now and we're we will start on  iteration of the impl and e2e testing it etc ... 

Thanks everybody!

jayunit100 commented:
bump :)

jayunit100 commented:
hiya folks, just checkin on this one - can we push it forward and iterate on the remaining details ?

jayunit100 commented:
/release-note-none

jayunit100 commented:
bump @liggitt  - still ok merge+iterate for this ? or looking for more before we go forward?

liggitt commented:
per https://github.com/kubernetes/enhancements/pull/2162#discussion_r553642761 this seems reasonable to merge as provisional

thockin commented:
/approve

Hit me with some followups and an impl PR!

jayunit100 commented:
/assign lavalamp

jayunit100 commented:
anyone got creds to approve this ?

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+evolution, but we don't need to do that as part of this KEP.
+
+It turns out that alternate [solutions](#alternatives):
+Break the old API in fundamental ways that are questionably WRT cost/benefit OR
```
I can't parse this sentence.

jayunit100 replied:
fixed

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+### Non-Goals
+
+* Publishing name labels on all resources - Not all resource names satisfy the label requirements. Some of those resources require
+a name that can be used as a DNS subdomain name.
```
Can you spell this out more. Is the sentence missing a "not" or something?

jayunit100 replied:
fixed

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
````diff
@@ -0,0 +1,487 @@
+If this label is missing, it is added by the apiserver.
+If this label is deleted, it is added by the apiserver.
+
+```
````
I don't really follow the diagram, like what what do all the lines represent?

jayunit100 replied:
fixed, its not needed really anymore

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+
+* The name of the namespace must satisfy label value [requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+Currently, the name of the namespace must be a valid [DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names),
+and this requirement must be satisfied going forward.
```
s/must be satisfied/will not change/

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is [reserved](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). However, since there
+is no validation of reserved label prefixes, an upgrade would mutate such a label.
```
given
* how long the label key is
* and its english meaning

I consider it extremely unlikely that someone would be using this label already *and* not desire that it have the value we want it to have.

jayunit100 replied:
yeah, agree.

jayunit100 replied:
pushed , thx

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+
+<<[UNRESOLVED Very likely to have a warning release, but not 100% sure ]>>
+
+- There is an ongoing debate, @thockin, @liggit, about wether a 'buffer' release is needed before this label is stomped.
```
It seems this is resolved now?

jayunit100 replied:
yes pushed

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+### Feature Enablement and Rollback
+
+1.21:
+- gate disabled by default
```
enable by default per other comment threads.

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+- release note
+
+1.23:
+- lock gate enabled
```
remove the gate. Also can this not be pushed up to 1.22?

jayunit100 replied:
yup, done pushing

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+Currently, the name of the namespace must be a valid [DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names),
+and this requirement must be satisfied going forward.
+
+### Risks and Mitigations
```
This is missing an important one: software built to assume the presence of the label is run on an old cluster which doesn't enforce the label.

jayunit100 replied:
fixed, pushing

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,487 @@
+* "It's kind of a special hack. The fact that it isn't generalized or generalizable 
+makes it feel special." - @thockin
+
+## Alternatives
```
Missing alternatives:
* add a virtual label. I.e., in memory only, don't store it on the object.
* Like a virtual label but go farther: add a syntax for label selectors that permit them to reference the fields in already in field selectors.

jayunit100 replied:
ah clever :) , yeah adding those

jayunit100 commented:
@lavalamp I think comments addressed
- removed the 3 version delayed strategy
- removed the diagram which I think is unnecessary at this point
- added the  "alternative" virtual in-memory label
- other sentence cleanups

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -172,14 +139,16 @@ NetworkPolicies or writing admission controller configurations.
 * The name of the namespace must satisfy label value [requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
 Currently, the name of the namespace must be a valid [DNS label](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names),
-and this requirement must be satisfied going forward.
+and this requirement will NOT CHANGE  going forward.
```
nit: all caps is maybe a bit aggressive :)

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -172,14 +139,16 @@ NetworkPolicies or writing admission controller configurations.
 fact that the whole "kubernetes.io" label prefix is [reserved](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set). However, since there
-is no validation of reserved label prefixes, an upgrade would mutate such a label.
+is no validation of reserved label prefixes, an upgrade would mutate such a label (this is not a major concern).
+* any clients/applications written to use this label will NOT be backwards compatible with older versions of the apiserver (which DONT enforce this label)
```
The worry is actually that they *would* be compatible (i.e., nothing obviously fails), and silently wouldn't have the desired security property.

jayunit100 replied:
in the conventional scenario, since policies are 'allow only', scheduling would **fail** because the namespace wouldn't have the requisite label, was what I meant, ill clarify that

lavalamp replied:
OK thanks.

lavalamp commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -408,6 +369,8 @@ makes it feel special." - @thockin
 ### Add new select-by-name capabilities to policy APIs
 
+* Add an 'in memory' virtual label that is defaulted by the apiserver but not stored in etcd
+* Add a new syntax for selectors on labels that allow them to dynamically select things in fields
```
Can you add reasons for / against? Like I'm not sure this last one is actually worse than the currently selected design.

thockin replied:
APIs are versioned.  Are we going to spec an API version in that syntax?  Are we going to auto-convert that field between API version?

This is the same problem that lead us to leave the various field-selector implementations a) hardcoded and b) VERY restricted.

abhiraut commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

@jayunit100 Looks like while squashing you added wrong email id on my behalf. it should be `Abhishek Raut <rauta@vmware.com>`

jayunit100 commented:
pushed the remaining fixes thanks again for the reviews today @lavalamp

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+functionalities offered by label selectors.
+
+There may also be many other usecases where the security or permissions boundary
+might be conveniently defined in terms of a namepsace name, and that such a
```
nit: namespace

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+- If this label is missing, it is added by the apiserver as a default.
+- If this label is deleted, it is added by the apiserver as a default.
+- defaults.go mutates namespaces on read, to always have the label `kubernetes.io/metadata.name=obj.Name`.
```
Using a defaulter, if someone tries to update it, would they get a success return code and then have their change not apply?  If so, could you clarify that behavior here.

jayunit100 replied:
yeah - you would never really be able to mutate the label.

jayunit100 replied:
ill updated the KEP to clarify this

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+NetworkPolicies allow users to open up traffic to/from their namespaces with the help of `namespaceSelector` field.
+However, there is a concern that arbitrary labels may not be the most secure way to open up traffic when it comes
+to namespace resource. In case a namespace's labels are known by others, any user can add labels at any time to any
```
normal users cannot add labels to a namespace.  Can you clarify what default role or clusterrole this user has?  Perhaps an explicit example would help.

jayunit100 replied:
sure, pushing this shortly

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+#### Simplicity
+
+It is being reported that users find it difficult to work with mutation and validation webhooks as they are not
+able to select a single namespace to enforce those controllers. Similarly, it is quite common and intuitive for
```
we recently had a conversation about this in apimachinery and concluded that trying to segment admission webhooks to a single namespace (or small numbers of namespaces) instead of excluding a small number of namespaces to avoid a bootstrapping loop was probably a bad idea because it uses a type system inconsistently across namespaces.

jayunit100 replied:
dded a comment to broaden this statement, pushing shortly

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+For example, you may need a way to declare namespaceAsName selectors, which is distinct from namespace label selectors.
+Introducing a new field has its own drawbacks: The obvious API, when combined with a podSelector, decreases security
+of a pod - an old client would be more open than users would expect them to be.
```
That depends on how you write the persisting logic.  One simple way to avoid is to say that when using this new field, a label selector is set by default which will never match anything.

jayunit100 replied:
yeah we discussed a few clever workarounds, they were though pretty odd.  ill add this comment as well , WRT workaround possibilities

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+namespace/pod combination, the policy implementer does something **much** more
+promiscuous.
+
+### Improve NetworkPolicy API and deprecate old fields
```
This seems worth exploring, with a discriminated union, you could extend this selection criteria in the future.

thockin replied:
Agree.  For this specific case we thought the more general solution described here would be best since it covers other users of NS selectors.

deads2k commented:
> > Looking over the proposal, you're trying to add a label selector that could only be used for `=`, `==`, `in` operations?
> 
> I think I could see `not in` being a differentiator in favor of creating a label like this.
> 
> Could you try to reduce the size of motivation to something more end user API focused and less focused on why you think a default label is the right solution? The information there could be in a different section about "why label selector"

Past me (https://github.com/kubernetes/enhancements/pull/2162#issuecomment-733740045)  re-convinced current me that this adds useful power.  I think clients may have some challenge trying to figure out when it is safe to use this, but obviously that eases over time.

I'm also ok with using defaulters.  I may have some interest in forcing the label to be empty or match a release ahead of beta.

@lavalamp I'm ok with this proposal, but it's novel enough to deserve an ack from you as well.

andrewsykim commented:
> Past me (#2162 (comment)) re-convinced current me that this adds useful power. I think clients may have some challenge trying to figure out when it is safe to use this, but obviously that eases over time.

I think we've also discussed building a light-weight backfill controller or admission webhook for older clusters that might want this functionality sooner. Not sure it's worth the trouble but definitely an option and there are folks willing to put the effort in.

lavalamp commented:
> I'm ok with this proposal, but it's novel enough to deserve an ack from you as well.

Yeah this isn't very pretty but the effort / reward ratio is there, so I'm ok with this.

jayunit100 commented:
hiya  @deads2k , finished addressing your comments.  With regard to exploring newer API structures, I didn't really do anything fo that comment because that is being explored in the sig-net-policy subproject already... hope that's ok :) ?  otherwise let me know if anything else is needed here 

thanks

deads2k commented:
> I think we've also discussed building a light-weight backfill controller or admission webhook for older clusters that might want this functionality sooner. Not sure it's worth the trouble but definitely an option and there are folks willing to put the effort in.

A backfill controller won't have the same guarantees that this mechanism provides.  In previous releases, it will still be possible to lie.

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -118,10 +118,9 @@ the namespace's name with traditional label selector mechanisms.
-namespace and thus send traffic to a pod belonging to this namespace. Matching a namespace by its name, on the other
-hand, is a more reliable way to whitelist namespaces as it's much easier to specifically allow only namespaces user
-control since user knows their own namespace names.
+to namespace resource. In case a namespace's labels are known by others, any user with wrote access to a namespace can add labels at any time to said namespace, and thus send traffic to a pod belonging to this namespace.
```
In the kube default policy, is there any user besides a cluster-admin who can do this?

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -136,6 +135,7 @@ able to select a single namespace to enforce those controllers. Similarly, it is
 users to remember the names of the namespaces rather than labels associated with those namespaces while defining
 NetworkPolicies or writing admission controller configurations.
 
+Thus, regardless of wether such a user story is valid, it is clearly 'simpler' to implement (as are other similar stories) when a default mechanism for namespace selection is available.
```
I think validity matters.  Things that aren't recommended don't need to be easy.

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -381,7 +381,7 @@ need for new API semantics in the networkpolicy specification, which is already
 For example, you may need a way to declare namespaceAsName selectors, which is distinct from namespace label selectors.
 Introducing a new field has its own drawbacks: The obvious API, when combined with a podSelector, decreases security
-of a pod - an old client would be more open than users would expect them to be.
+of a pod - an old client would be more open than users would expect them to be (in the most obvious implementations).  We note that there are workarounds involving special "always off" selectors, which can avoid this scenario, but they come with their own obvious inherent technical debt and costs.
```
> own obvious inherent technical debt and costs

this is fairly opinionated.  I don't know that I see significant technical debt here.

deads2k commented:
I don't agree with the clearly's and obviously's, but I can see value here in specifying not-in values.


/lgtm
/approve

/hold

holding for a squash and to give @thockin another shot at review if he'd like it.  if I forget, remind me to release the hold in a week or so.

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+### Feature Enablement and Rollback
+
+1.21:
```
alphas are off by default.

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+* **How can an operator determine if the feature is in use by workloads?**
+
+By running `kubectl get ns --show-labels` and inspecting the `kubernetes.io` values.
```
this isn't quite right.  You'd need to track list/watch requests keying against this name.

jayunit100 commented:
thanks @deads2k ! is it Ok to keep 3 commits separate since there were different authors (me, Ricardo, abhishek) (and I squash the 4th) ?

deads2k commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine** 
+
+Since this just an immutable field, we don't expect any such indicators to be of relevance.
```
I think a metric adding the number of rejections for invalid use of the label is fairly reasonable.  That would be unexpected client behavior even though it ought not happen.

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+is no validation of reserved label prefixes, an upgrade would mutate such a label (this is not a major concern).
+* any clients/applications expecting this label (i.e. for a NetworkPolicy), wouldnt see it, and thus they might not be able to find namespaces using this label technique.  In the case of network policies, this would result in traffic being blocked due to unavailable namespace labels.
+
+For this reason, we can do one alpha release where we log/event if we observe usage of this key that is not compatible.
```
Are we going to do an alpha or not?  I thought not.

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+<<[UNRESOLVED See https://github.com/kubernetes/kubernetes/pull/96968 where this evolves, but this example is functionally valid, modulo warnings ]>>
```
My understanding of the discussion is that this is resolved.  Setting the value should be unconditional.  It's our label. No need to preserve the value.

The only open question is whether to do it silently in defaults or with a warning in strategy.  The warning seems appropriate to me, but we'll need to cover all create/update/get paths, which is slightly more complicated than defaults.go

If we don't do the warning, we can do it in defaults and just stomp the value.

I'm advocating for the warning, but only weakly.

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/README.md:
```diff
@@ -0,0 +1,452 @@
+
+1.22:
+- gate enabled by default
+- remove gate
```
strictly speaking this would lock the gate, and some releases later we would remove it.

thockin reviewed (commented):
I am good with this as-is.  Some small followups and it can go to implementable.

/remove hold

thockin commented on keps/sig-api-machinery/2161-apiserver-default-labels/kep.yaml:
```diff
@@ -0,0 +1,33 @@
+replaces:
+stage: alpha
+latest-milestone: "v1.21"
+milestone:
```
this needs to be updated
